### ISSUE TITLE
Filter with time component

### ISSUE THREAD

**Issue by daentech** [wysaid/android-gpuimage-plus#89]
Hi,

I'm looking to add a native video filter which changes over time, in the same way as the DynamicWaveFilter. Can I achieve this with the library? Can other uniforms be provided to the shaders?

Thanks in advance, and sorry for asking in English :)

```
varying mediump vec2 textureCoordinate;
uniform sampler2D inputImageTexture;
uniform float iGlobalTime;

uniform vec2 noDraws;

void main()
{
    vec2 position = textureCoordinate * 256.0;

    vec2 uv = 0.5 + (position-0.5)*(0.9 + 0.1*sin(0.2*iGlobalTime));

    vec3 oricol = texture2D( inputImageTexture, textureCoordinate).rgb;
    vec3 col;

    col.r = texture2D(inputImageTexture,vec2(uv.x+0.003,-uv.y)).r;
    col.g = texture2D(inputImageTexture,vec2(uv.x+0.000,-uv.y)).g;
    col.b = texture2D(inputImageTexture,vec2(uv.x-0.003,-uv.y)).b;

    col = clamp(col*0.5+0.5*col*col*1.2,0.0,1.0);

    col *= 0.5 + 0.5*16.0*uv.x*uv.y*(1.0-uv.x)*(1.0-uv.y);

    col *= vec3(0.95,1.05,0.95);

    col *= 0.9+0.1*sin(10.0*iGlobalTime+uv.y*1000.0);

    col *= 0.99+0.01*sin(110.0*iGlobalTime);

    col = mix( col, oricol, clamp(-2.0+2.0*position.x,0.0,1.0) );

    gl_FragColor.bgra = vec4(col.b, col.g, col.r,1.0);
}
```

**Comment by wysaid** [comment]
Of course, you can try it.
PS: The whole repo will be open source in a few days/weeks.
You can add more custom functions at that time.😋

**Comment by daentech** [comment]
Great :) Is this achievable at the moment? I can't see how to provide iGlobalTime or noDraws to the shader

**Comment by wysaid** [comment]
Yes, you can override the 'render2texture' function, and pass these values every frame.
Or you have to write a wrapper for java-layer, and pass the uniforms as you wish.

**Comment by daentech** [comment]
Thanks, I'll give it a try 👍

**Comment by chimanaco** [comment]
Could you please provide an example code for this?
For instance, how can I send a value from a seekbar to a custom filter as uniform?

**Comment by wysaid** [comment]
@chimanaco

**Comment by chimanaco** [comment]
Thank you. I tried it but I haven't gotten luck yet. The below is the flow I followed and I overrided the `setIntensity` function in CustomFilter_0. I just want to pass a value without changing an intensity of a filter itself.

1. CameraDemoActivity.Java
`
public void onProgressChanged(SeekBar seekBar, int progress, boolean fromUser) {
                float intensity = progress / 100.0f;
                mCameraView.setFilterIntensity(intensity);
            }`

2. CameraGLSurfaceView.Java
`public void setFilterIntensity(final float intensity) {
        queueEvent(new Runnable() {
            @Override
            public void run() {
                if (mFrameRecorder != null) {
                    mFrameRecorder.setFilterIntensity(intensity);
                } else {
                    Log.e(LOG_TAG, "setFilterIntensity after release!!");
                }
            }
        });
    }`

3. CGEFrameRenderer.Java
`//set the intensity of the filter
    public void setFilterIntensity(float value) {
        if(mNativeAddress != 0)
            nativeSetFilterIntensity(mNativeAddress, value);
    }`

4. cgeFrameRendererWrapper.cpp
`JNIEXPORT void JNICALL Java_org_wysaid_nativePort_CGEFrameRenderer_nativeSetFilterIntensity
  (JNIEnv *env, jobject, jlong addr, jfloat intensity)
{
	CGEFrameRenderer* renderer = (CGEFrameRenderer*)addr;
	renderer->setFilterIntensity(intensity);
}`

5. cgeFrameRenderer.cpp
`void CGEFrameRenderer::setFilterIntensity(float value)
	{
		CHECK_RENDERER_STATUS;

		m_resultMutex.lock();
		auto& filters = m_frameHandler->peekFilters();
		for(auto& filter : filters)
		{
			filter->setIntensity(value);
		}
		m_resultMutex.unlock();
	}`

6. CustomFilter_0.cpp
`
#include "customFilter_0.h"

using namespace CGE;

static CGEConstString s_fsh = CGE_SHADER_STRING_PRECISION_H
(
 varying vec2 textureCoordinate;
 uniform sampler2D inputImageTexture;
 uniform vec2 vSteps;
 uniform float intensity;
  
 void main(void)
{
    vec4 originalColor = texture2D(inputImageTexture, textureCoordinate);
    vec4 invertedColor = vec4(((intensity * 0.5+0.5) - originalColor.rgb), originalColor.w);
    gl_FragColor = invertedColor;

});

bool CustomFilter_0::init()
{
    if(m_program.initWithShaderStrings(g_vshDefaultWithoutTexCoord, s_fsh))
    {
        m_program.bind();
        mStepLoc = m_program.uniformLocation("vSteps");
        return true;
    }
    return false;
}

void CustomFilter_0::render2Texture(CGEImageHandlerInterface* handler, GLuint srcTexture, GLuint vertexBufferID)
{
    handler->setAsTarget();
    m_program.bind();
    
    const CGESizei& sz = handler->getOutputFBOSize();
    
    glUniform2f(mStepLoc, 1.0f / sz.width, 1.0f / sz.height);
    
    glEnableVertexAttribArray(0);
    glVertexAttribPointer(0, 2, GL_FLOAT, GL_FALSE, 0, 0);
    glActiveTexture(GL_TEXTURE0);
    glBindTexture(GL_TEXTURE_2D, srcTexture);

    glDrawArrays(GL_TRIANGLE_FAN, 0, 4);
}

CGEConstString CustomFilter_0::paramIntensity = "intensity";

void CustomFilter_0::setIntensity(float value)
	{
		m_program.bind();
		m_program.sendUniformf(paramIntensity, value);
	}
`
Thanks in advance.

**Comment by wysaid** [comment]
You can make a fork and push your code.
I'll try your code.

**Comment by chimanaco** [comment]
Ok. This is my test code.
https://github.com/chimanaco/android-gpuimage-plus/tree/features/customfilter

After building the native library, go to `CameraDemoActivity`, find `Custom` and tap it. You will see an inverted image.
What I expect is changing value instead of intensity of filter itself. 

Thanks in advance!

**Comment by wysaid** [comment]
Please see the [PR](https://github.com/chimanaco/android-gpuimage-plus/pull/1)
You can sync the latest code, and pass false for arg `useWrapper`
